### PR TITLE
refactor: box-based context

### PR DESCRIPTION
**Pull request by enitrat**
Wraps StaticContext and DynamicContext in boxes, reducing the amount of elements re-written to memory on each struct modification.
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: #43 

## What is the new behavior?
StaticExecutionContext and DynamicExecutionContext are now boxxed.

- Added relevant methods on ExecutionContext to access the fields required.
- A side effect is that reads on DyamicExecutionContext fields might be more expensive, because we have to pass it as ref due to ownership mechanism
-

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Added changes to [`CHANGELOG.md`](../CHANGELOG.md)

- [x] Yes
- [ ] No

## Other information

Gas info

```shell
___IMPROVEMENTS___
kakarot::tests::test_execution_context::test_execution_context_read_code 84010 --> 79810 | -5.00 %
kakarot::tests::test_execution_context::test_is_caller_eoa 94210 --> 91910 | -2.44 %
kakarot::tests::test_instructions::test_comparison_operations::test_and_max_and_max 255038 --> 250638 | -1.73 %
kakarot::tests::test_instructions::test_comparison_operations::test_and_two_random_uint 255038 --> 250638 | -1.73 %
kakarot::tests::test_instructions::test_comparison_operations::test_and_zero_and_max 255038 --> 250638 | -1.73 %
kakarot::tests::test_instructions::test_comparison_operations::test_not_max_uint 155660 --> 151260 | -2.83 %
kakarot::tests::test_instructions::test_comparison_operations::test_not_random_uint 155660 --> 151260 | -2.83 %
kakarot::tests::test_instructions::test_comparison_operations::test_not_zero 155660 --> 151260 | -2.83 %
kakarot::tests::test_instructions::test_comparison_operations::test_xor_different_pair 255038 --> 250638 | -1.73 %
kakarot::tests::test_instructions::test_comparison_operations::test_xor_half_same_pair 255038 --> 250638 | -1.73 %
kakarot::tests::test_instructions::test_comparison_operations::test_xor_same_pair 255038 --> 250638 | -1.73 %
kakarot::tests::test_instructions::test_duplication_operations::test_dup1 250130 --> 242130 | -3.20 %
kakarot::tests::test_instructions::test_duplication_operations::test_dup10 478170 --> 470170 | -1.67 %
kakarot::tests::test_instructions::test_duplication_operations::test_dup11 503030 --> 495030 | -1.59 %
kakarot::tests::test_instructions::test_duplication_operations::test_dup12 527890 --> 519890 | -1.52 %
kakarot::tests::test_instructions::test_duplication_operations::test_dup13 552750 --> 544750 | -1.45 %
kakarot::tests::test_instructions::test_duplication_operations::test_dup14 577610 --> 569610 | -1.39 %
kakarot::tests::test_instructions::test_duplication_operations::test_dup15 602470 --> 594470 | -1.33 %
kakarot::tests::test_instructions::test_duplication_operations::test_dup16 627330 --> 619330 | -1.28 %
kakarot::tests::test_instructions::test_duplication_operations::test_dup2 279290 --> 271290 | -2.86 %
kakarot::tests::test_instructions::test_duplication_operations::test_dup3 304150 --> 296150 | -2.63 %
kakarot::tests::test_instructions::test_duplication_operations::test_dup4 329010 --> 321010 | -2.43 %
kakarot::tests::test_instructions::test_duplication_operations::test_dup5 353870 --> 345870 | -2.26 %
kakarot::tests::test_instructions::test_duplication_operations::test_dup6 378730 --> 370730 | -2.11 %
kakarot::tests::test_instructions::test_duplication_operations::test_dup7 403590 --> 395590 | -1.98 %
kakarot::tests::test_instructions::test_duplication_operations::test_dup8 428450 --> 420450 | -1.87 %
kakarot::tests::test_instructions::test_duplication_operations::test_dup9 453310 --> 445310 | -1.76 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test__exec_sdiv_neg 352170 --> 347770 | -1.25 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_add 298650 --> 294250 | -1.47 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_add_overflow 256180 --> 251780 | -1.72 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_addmod 339030 --> 334630 | -1.30 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_addmod_by_zero 339030 --> 334630 | -1.30 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_addmod_overflow 339030 --> 334630 | -1.30 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_div 265620 --> 261220 | -1.66 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_div_by_zero 265620 --> 261220 | -1.66 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_exp 347410 --> 343010 | -1.27 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_exp_overflow 354360 --> 349960 | -1.24 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_mod 265620 --> 261220 | -1.66 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_mod_by_zero 345740 --> 341340 | -1.27 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_mul 270540 --> 266140 | -1.63 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_mul_overflow 276970 --> 272570 | -1.59 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_sdiv_by_0 345740 --> 341340 | -1.27 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_sdiv_pos 345740 --> 341340 | -1.27 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_signextend 594046 --> 589646 | -0.74 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_signextend_no_effect 594046 --> 589646 | -0.74 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_signextend_on_negative 868366 --> 863966 | -0.51 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_smod 345740 --> 341340 | -1.27 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_smod_neg 345740 --> 341340 | -1.27 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_smod_zero 265620 --> 261220 | -1.66 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_stop 60310 --> 58010 | -3.81 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_sub 256180 --> 251780 | -1.72 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_exec_sub_underflow 256180 --> 251780 | -1.72 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_mulmod_basic 364430 --> 360030 | -1.21 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_mulmod_overflow 364430 --> 360030 | -1.21 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_mulmod_zero 364430 --> 360030 | -1.21 %
kakarot::tests::test_instructions::test_stop_and_arithmetic_operations::test_mulmod_zero_modulus 364430 --> 360030 | -1.21 %
kakarot::tests::test_kakarot::nominal_case_empty_pc 3931772 --> 2932972 | -25.40 %



___WORSENED___
kakarot::tests::test_execution_context::test_execution_context_new 191040 --> 207740 8.74 %
kakarot::tests::test_execution_context::test_execution_context_revert 113870 --> 117570 3.25 %
kakarot::tests::test_execution_context::test_execution_context_stop_and_revert 55310 --> 56710 2.53 %
kakarot::tests::test_stack::pop::test_pop_return_err_when_stack_underflow 33440 --> 37140 11.06 %
Overall gas change: performance improvement, gas consumption-1.03 %
```

### CODE REVIEW

File: scripts/compare_snapshot.py [hunk 1]
```diff
@@ -19,7 +19,7 @@ def compare_snapshots(current, previous):
     """Compare current and previous snapshots and return differences."""
```
Comment (Eikix) [inline scripts/compare_snapshot.py:19]: what was the result of compare_snapshot?

Comment (enitrat) [reply scripts/compare_snapshot.py:19]: It's in the description

File: src/context.cairo [hunk 2]
```diff
@@ -192,98 +208,126 @@ impl ExecutionContextImpl of ExecutionContextTrait {
...
-    fn read_code(ref self: ExecutionContext, len: usize) -> Span<u8> {
-        // Copy code slice from [pc, pc+len]
-        let code = self.static_context.call_context.bytecode().slice(self.program_counter, len);
-
-        self.program_counter += len;
-        code
+    fn stopped(ref self: ExecutionContext) -> bool {
+        let dyn_ctx = self.dynamic_context.unbox();
+        let stopped = dyn_ctx.stopped;
+        self.dynamic_context = BoxTrait::new(dyn_ctx);
```
Comment (Eikix) [inline src/context.cairo:227]: why is this needed? can't we just read what's inside the box without needing to destruct and reconstruct?

 whats innn the boxxxx (movie joke)

File: src/context.cairo [hunk 3]
```diff
@@ -192,98 +208,126 @@ impl ExecutionContextImpl of ExecutionContextTrait {
...
+        return_data
     }
 
+    /// Stops the current execution context.
     #[inline(always)]
-    fn events(self: @ExecutionContext) -> Span<Event> {
-        self.dynamic_context.events.span()
+    fn stop(ref self: ExecutionContext) {
+        let mut dyn_ctx = self.dynamic_context.unbox();
+        dyn_ctx.stopped = true;
```
Comment (Eikix) [inline src/context.cairo:272]: here I understand why  we unbox, we're writing
but above, we're just reading, so i'm confused

File: src/context.cairo [hunk 4]
```diff
@@ -192,98 +208,126 @@ impl ExecutionContextImpl of ExecutionContextTrait {
...
+        // Copy code slice from [pc, pc+len]
+        let code = self
+            .static_context
+            .unbox()
+            .call_context
+            .bytecode()
+            .slice(self.program_counter, len);
+
+        self.program_counter += len;
+        code
```
Comment (Eikix) [inline src/context.cairo:330]: here, we're not "reboxing"?

**Review by Eikix — COMMENTED** [review]
just a few questions

File: scripts/compare_snapshot.py (hunk 1 above)
Comment (Eikix) [reply scripts/compare_snapshot.py:19]: swag

File: src/context.cairo [hunk 5]
```diff
@@ -22,7 +22,12 @@ struct CallContext {
     value: u256,
 }
 
-// I think we should not directly access the fields of the call context;
+
+// *************************************************************************
+//                              CallContext
+// *************************************************************************
```
Comment (danilowhk) [inline src/context.cairo:28]: These are nice!

File: src/context.cairo (hunk 4 above)
Comment (danilowhk) [reply src/context.cairo:330]: Also curious as its seems we are not reboxing on evm_address, starknet_address and read_only

File: src/context.cairo (hunk 2 above)
Comment (enitrat) [reply src/context.cairo:227]: Because `DynamicExecutionContext` doesn't implement Copy, we can't do the same thing we're doing with `StaticExecutionContext`. 

Imagine if we could - then we would be able to unbox `DynamicExecutionContext` in two different variables, and we could append to its arrays in two different locations.
So, our only solution is to pass it as ref. 

`let dyn_ctx = self.dynamic_context.unbox();` takes ownership of the dynamic context, therefore as we need to return it, we need to update the struct field to a new pointer

File: src/context.cairo (hunk 3 above)
Comment (enitrat) [reply src/context.cairo:272]: same reasoning as above except that in this one we modify the value

File: src/context.cairo (hunk 4 above)
Comment (enitrat) [reply src/context.cairo:330]: static context is copyable

File: src/context.cairo (hunk 5 above)
Comment (enitrat) [reply src/context.cairo:28]: yeah, big files tend to get a bit hard to navigate in, I was thinking it would be helpful here

File: src/context.cairo (hunk 2 above)
Comment (Eikix) [reply src/context.cairo:227]: we are not allowed to access the value of a box directly? since it's just a read, like in pseudocode:
`let stopped = self.dynamic_context.stopped`

File: src/context.cairo (hunk 4 above)
Comment (Eikix) [reply src/context.cairo:330]: so calling unbox() will copy static_context?

File: src/context.cairo (hunk 2 above)
Comment (enitrat) [reply src/context.cairo:227]: no, you can only have one ownership of a pointer at a time

File: src/context.cairo (hunk 4 above)
Comment (enitrat) [reply src/context.cairo:330]: yes but it's a "copy" only at the type level

File: src/context.cairo (hunk 2 above)
Comment (Eikix) [reply src/context.cairo:227]: how about `self.dynamic_context.stopped.clone()` or `self.dynamic_context.stopped.copy()`

Comment (Eikix) [reply src/context.cairo:227]: it just feels bad that we can't deref a box's member here like in rust

**Review by Eikix — APPROVED** [review]
lgtm 🔥🔥🔥🔥🔥🔥🔥
